server/server.go

```diff
@@ -1930,9 +1952,34 @@ func (s *Server) createClient(conn net.Conn, ws *websocket) *client {
 	// Re-Grab lock
 	c.mu.Lock()
 
+	tlsRequired := ws == nil && info.TLSRequired
+	var pre []byte
+	// If we have both TLS and non-TLS allowed we need to see which
+	// one the client wants.
+	if opts.TLSConfig != nil && opts.AllowNonTLS {
+		pre = make([]byte, 8)
+		c.nc.SetReadDeadline(time.Now().Add(25 * time.Millisecond))
+		n, err := c.nc.Read(pre[:])
+		c.nc.SetReadDeadline(time.Time{})
+		pre = pre[:n]
+		// Assume TLS unless we see nothing or CONNECT.
+		if err != nil || bytes.Contains(pre, []byte("CONNECT")) {
+			tlsRequired = false
+		} else {
+			tlsRequired = true
+		}
+	}
+
 	// Check for TLS
 	if tlsRequired {
 		c.Debugf("Starting TLS client connection handshake")
+		// If we have a prebuffer create a multi-reader.
+		if len(pre) > 0 {
+			c.nc = &tlsMixConn{c.nc, bytes.NewBuffer(pre)}
```

Comment:
Do we really need that? If you create a buffer (not a bufio that we don't know how much it is going to be read), then you know exactly how many bytes have been read from the socket. So you could simply pass to the readloop the "pre" buffer and call parse on it as you do, then resume with reading from socket with readLoop's buf.

Reply:
Not for TLS..

Reply:
Of course, wasn't thinking straight.